designs/0003-ragged-containers.md

```diff
@@ -0,0 +1,575 @@
+to match if the variable being assigned is a sized block variable.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+It will take developer effort to code, test, document, and maintain.
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+A [preliminary
+design](https://github.com/stan-dev/stan/wiki/Ragged-array-spec) was
+refined online and motivated this design.  I don't consider it a
+serious alternative to this proposal.
+
+# Prior art
+[prior-art]: #prior-art
+
+For now, what we do is code things in one of two long forms.  Suppose
+we have a sequence of sequences, each of which we want to model as a
+time-series, say
+
+```
+int K;             // num items
+int<lower = 0> N;  // observations
+real[N] y;         // num observations
+int[K] start;      // start of each series
+int[K] end;        // end of each series
+...
+for (k in 1:K)
+  y[start[k], end[k]] ~ foo(theta);
+```
+
+For example, we might have data
+
+```
+N = 15;
+K = 4;
+start = {1, 5, 7, 12};
+end = {5, 7, 12, 16};
+y = {1, 2, 3, 0, -1, 0.2, 3, 5,
+     7, 9, 11, 15, 13, 11, 9};
+```
+
+With the new proposal, this will look like:
+
+```
+real[{4, 2, 5, 9}] yr
+  = {{1, 2, 3, 0}, {-1, 0.2},
+     {3, 5, 7, 9, 11}, {15, 13, 11, 9 }};
+...
+for (k in 1:K)
+  yr[k] ~ foo(theta);
+```
+
+The new data structure will also be more efficient as a reference can
+be returned from `yr[k]` rather than requiring a copy.
+
+In some situations, long form can be used if there's no inherent
+structure in the subarrays.  For example, we might have long form data
+
+```
+real[N] y
+  = {1, 2, 3, 0, -1, 0.2, 3, 5,
+     7, 9, 11, 15, 13, 11, 9};
+int<lower = 1, upper = K>[N] kk = {1, 1, 1, 1, 2, 2, 3, 3, 3,
+              3, 3, 4, 4, 4, 4};
+...
+for (n in 1:N)
+  y[n] ~ foo(theta[kk[n]]);
+```
+
+whereas the ragged structure would either be
+
+```
+for (k in 1:K)
+  y[k] ~ foo(theta[k]);
+```
+
+or even
+
+```
+y ~ foo(theta);
+```
+
+if `foo` is vectorized.
+
+### Other languages
+
+In R, they use lists, which allow heterogeneous containers.   The
+ragged proposal here is more strongly typed.  MATLAB does not support
+ragged arrays.
+
+A language like C++ supports ragged arrays in the sense of allowing
+containers to hold containers.  For example, you could have
+
+```
+using std::vector;
+vector<vector<double>> y;
+for (size_t i = 0; i < 100; ++i)
+  y[i] = vector(0.0, i);  // y[i].size() == i
+```
+
+This C++ idiom will be supported when we have unsized local variables.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Whether this gets implemented for the deprecated Rdump format of I/O.
+(The JSON format should not need to change other than that loops need
+to be size-bounded rather than constant-bounded.)
+
+Ragged containers will play nicely with unsized local variables and
+function arguments.
+
+This proposal as written depends on allowing type declarations like:
```

Comment:
This was implemented in Stan 2.25 where we now have array[...] real x, array[...] vector[N] a, etc. so we can remove the text below.